### ISSUE TITLE
Local functions don't give unreachability diagnostics

### ISSUE THREAD

**Issue by RikkiGibson** [dotnet/roslyn#58304]
[SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4AEBMBGAsAKAIBkB7AYwgBsAKASgG4D1sAGAAgA9H8OA6AFRIBlAC4AnWAHM69NgHo5bGCTYBTMWJJiCTACxtSFGrTYBeAHxsRAC00B3JaocBRDmVUAHEVBIwZQA)

```cs
using System;

Local();
string x;
x.ToString(); // no error

void Local() => throw new Exception();
```

Local functions can propagate out unreachable state in definite assignment. The lack of any warning or error diagnostics on `string x`; or `x.ToString();` ends up being a little misleading. Specifically @jcouv and I just spent several minutes trying to identify a definite assignment issue which was actually by-design :)

It feels like we should give an unreachability warning for the statements following the local function call.

**Comment by jaredpar** [comment]
That sounds reasonable but wanted to let @agocke chime in here in case there are any subtle issues that we are missing.

**Comment by agocke** [comment]
I don't see a problem with changing this, but FYI it was kind of by-design. The background here is that in both the spec and the compiler, definite assignment and control flow are separate. It turns out that data flow is more precise than the control flow spec, so you can end up with cases where definite assignment says a line is unreachable, but control flow has no idea.

For instance, this program produces no warnings or errors:

```C#
        const bool x = true;
        int y;
        if ("".Length == 0 || x)
            throw new Exception();
        y++;
```

You have two options here: you could patch up this case in the control flow pass, or you could try to unify the two passes. Either way, you'll probably end up with some spec work here to say what the new behavior is. I don't think I really modified the control flow pass at all, so the current behavior is probably almost identical to C# 6

**Comment by jaredpar** [comment]
Closing as this behavior is effectively by design as it's unreachable code.
